eth/stagedsync/stage_hashstate.go

```diff
@@ -169,34 +151,126 @@ func PromoteHashedStateCleanly(logPrefix string, db kv.RwTx, cfg HashStateCfg, q
-	return func(k, v []byte, next etl.ExtractNextFunc) error {
-		newK, err := transformKey(k)
+func readPlainStateOnce(
+	logPrefix string,
+	db kv.RwTx,
+	fromBucket string,
+	tmpdir string,
+	loadFunc etl.LoadFunc,
+	quit <-chan struct{},
+) error {
+	bufferSize := etl.BufferOptimalSize
+	buffer := etl.NewSortableBuffer(bufferSize)
+
+	collector1 := etl.NewCollector(logPrefix, tmpdir, buffer)
+	collector2 := etl.NewCollector(logPrefix, tmpdir, buffer)
+	defer func() {
+		collector1.Close()
+		collector2.Close()
+	}()
+
+	t := time.Now()
+	logEvery := time.NewTicker(30 * time.Second)
+	defer logEvery.Stop()
+	var m runtime.MemStats
+
+	c, err := db.Cursor(fromBucket)
+	if err != nil {
+		return err
+	}
+	defer c.Close()
+
+	convertAccFunc := func(key []byte) ([]byte, error) {
+		hash, err := common.HashData(key)
+		return hash[:], err
+	}
+
+	convertStorageFunc := func(key []byte) ([]byte, error) {
+		addrHash, err := common.HashData(key[:length.Addr])
 		if err != nil {
-			return err
+			return nil, err
 		}
-		return next(k, newK, v)
+		inc := binary.BigEndian.Uint64(key[length.Addr:])
+		secKey, err := common.HashData(key[length.Addr+length.Incarnation:])
+		if err != nil {
+			return nil, err
+		}
+		compositeKey := dbutils.GenerateCompositeStorageKey(addrHash, inc, secKey)
+		return compositeKey, nil
 	}
-}
 
-func keyTransformExtractAcc(transformKey func([]byte) ([]byte, error)) etl.ExtractFunc {
-	return func(k, v []byte, next etl.ExtractNextFunc) error {
-		if len(k) != 20 {
-			return nil
+	var startkey, endkey []byte
+
+	// reading kv.PlainState
+	for k, v, e := c.Seek(startkey); k != nil; k, v, e = c.Next() {
+		if e != nil {
+			return e
 		}
-		newK, err := transformKey(k)
-		if err != nil {
+		if err := libcommon.Stopped(quit); err != nil {
 			return err
 		}
-		return next(k, newK, v)
-	}
-}
 
-func keyTransformExtractStorage(transformKey func([]byte) ([]byte, error)) etl.ExtractFunc {
-	return func(k, v []byte, next etl.ExtractNextFunc) error {
 		if len(k) == 20 {
+			newK, err := convertAccFunc(k)
+			if err != nil {
+				return err
+			}
+			if err := collector1.Collect(newK, v); err != nil {
+				return err
+			}
+		} else {
+			newK, err := convertStorageFunc(k)
+			if err != nil {
+				return err
+			}
+			if err := collector2.Collect(newK, v); err != nil {
+				return err
+			}
+		}
+
+		select {
+		default:
+		case <-logEvery.C:
+			logArs := []interface{}{"from", fromBucket}
+			var additionalLogArguments etl.AdditionalLogArguments
+			if additionalLogArguments != nil {
+				logArs = append(logArs, additionalLogArguments(k, v)...)
+			} else {
+				logArs = append(logArs, "current key", fmt.Sprintf("%x...", k[:6]))
+			}
+
+			runtime.ReadMemStats(&m)
+			logArs = append(logArs, "alloc", libcommon.ByteCount(m.Alloc), "sys", libcommon.ByteCount(m.Sys))
+			log.Info(fmt.Sprintf("[%s] ETL [1/2] Extracting", logPrefix), logArs...)
+		}
+		if endkey != nil && bytes.Compare(k, endkey) > 0 {
```

Comment:
endkey is always nil. 
(startKey also - can use cursor.First() then)

Reply:
I guess I can replace `if endkey != nil && bytes.Compare(k, endkey) > 0 {` with `if k != nil{`, can't I?

Reply:
`k != nil` already exists in `for` loop condition. Can remove this line.